Increase task wdt timeout for ESP32/ESP32-C3

agners opened the pull request:
# What does this implement/fix? 

Introduce application_esp32.cpp/application_esp8266.cpp and 
pre_/post_setup_arch_(). Use the function to increase the Task 
Watchdog timeout from 5s to 10s before entering `setup()` phase.

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes esphome/issues#2277

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [x] ESP32/ESP32-C3
- [ ] ESP8266

## Example entry for `config.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example config.yaml

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

agners committed:
Separate architecture specific code

Introduce application_esp32.cpp/application_esp8266.cpp which allows
to separate architecture specific code.

agners committed:
Increase task watchdog timeout during setup()

Introduce pre_/post_setup_arch_(). Use the function to increase the
Task Watchdog timeout from 5s to 10s.

jesserockz reviewed (approved):
LGTM

OttoWinter commented on esphome/core/application_esp32.cpp:
```diff
@@ -0,0 +1,21 @@
+  esp_task_wdt_init(30, true);
+}
+
+void Application::post_setup_arch_() { esp_task_wdt_init(5, true); }
```
So if I'm reading [this](https://github.com/espressif/arduino-esp32/blob/master/cores/esp32/main.cpp#L54) correctly, the "normal" ESP32 does not currently have any WDT enabled.

5 seconds seems like plenty, so it shouldn't cause any issues other than maybe badly written custom code. Just something we should track if we see issues after the next release.

agners replied:
> So if I'm reading this correctly

I don't think that this enables or disables the Task WDT. It seems that just forces a watchdog feed before entering `loop()`.  Afaict, the normal ESP32 _does_ have Task WDT enabled with the same 5s timeout (at least according to the sdkconfig in the Arduino SDK, see link in https://github.com/esphome/issues/issues/2277). 

> 5 seconds seems like plenty

That is the time all setup tasks together seem to take currently, on single core. From the log it seems that most time is spent in WiFi connect. My guess is that the single core variant busy waits in that area which does not cause the idle task to get scheduled. In the dual core variant, the WiFi code runs on the first core and Arduino/ESPHome on the second. It seems that in this case the second CPU is put asleep while connecting to WiFi, which causes the idle task to trigger the Task Watchdog. That is at least how I understand the difference.

Now that I think a bit more about it, maybe it would be smarter to just call `esp_task_wdt_reset()` in `Application::feed_wdt()` instead of increasing the timeout. That way the task watchdog still would trigger in case some setup does something which takes more than 5s.

OttoWinter replied:
> That is the time all setup tasks together seem to take currently, on single core. From the log it seems that most time is spent in WiFi connect. My guess is that the single core variant busy waits in that area which does not cause the idle task to get scheduled. In the dual core variant, the WiFi code runs on the first core and Arduino/ESPHome on the second. It seems that in this case the second CPU is put asleep while connecting to WiFi, which causes the idle task to trigger the Task Watchdog. That is at least how I understand the difference.

Ah that makes sense

> Now that I think a bit more about it, maybe it would be smarter to just call esp_task_wdt_reset() in Application::feed_wdt() instead of increasing the timeout. That way the task watchdog still would trigger in case some setup does something which takes more than 5s.

If the issue is with wifi, I don't think this will help? Because then the code being executed is all in esp-idf/arduino-esp32 - `feed_wdt` is never called, so the reset wouldn't happen.

But if this is only needed on single-core variants, is there a macro/function we can check to only enable this on those variants?

agners replied:
> If the issue is with wifi, I don't think this will help? Because then the code being executed is all in esp-idf/arduino-esp32 - feed_wdt is never called, so the reset wouldn't happen.

The code does return from the WiFi calls. The `can_proceed()` function gets called until the task WDT triggers. See this log output: https://pastebin.com/b1QQCHm3 (I did set the task WDT timeout to 2s, but not really relevant in this case). Afaict, the processing is deferred to another FreeRTOS task. The problem is that just calling `yield()` does not lead feed the watchdog since it won't schedule the idle task. This Redditor experienced [the same](https://www.reddit.com/r/esp32/comments/e52b4m/why_did_changing_from_yield_to_delay_fix_a/).

TBH, I am not exactly clear what triggers the idle task on the dual-core ESP32s. I only guess that the inter-processor communication somehow causes the task to sleep/block which makes the CPU available for the idle task.


I tried calling `esp_task_wdt_reset()` in `Application::feed_wdt()` but it returns `-261` which is `ESP_ERR_NOT_FOUND`. The [documentation says](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/system/wdts.html#_CPPv418esp_task_wdt_resetv) this means "the current running task has not subscribed to the TWDT". I guess that means that reset only really can be called from the task which initialized the Task WDT, in this case which is the idle task (see [components/freertos/port/port_common.c](https://github.com/espressif/esp-idf/blob/release/v4.3/components/freertos/port/port_common.c#L118)). So what `CONFIG_ESP_TASK_WDT_CHECK_IDLE_TASK_CPU[X]` is really doing is making sure the CPU never stays busy for more than 5s at a time.

What we could do is to insert a `delay(1)` instead of [this](https://github.com/esphome/esphome/blob/v1.20.3/esphome/core/application.cpp#L55) `yield()` while continue processing the startup tasks. This makes sure that idle gets scheduled and the watchdog feed.

Or we indeed just tell the task watchdog that a higher "idle task starvation" is ok during startup phase by setting the timeout to e.g. 10s (what the code is currently doing).

> But if this is only needed on single-core variants, is there a macro/function we can check to only enable this on those variants?

That should be doable, the config pre-processor variables are available afict. E.g. `CONFIG_FREERTOS_UNICORE`.

OttoWinter replied:
Interesting! I guess it makes sense `yield()` doesn't schedule the idle task.

> What we could do is to insert a delay(1) instead of this yield() while continue processing the startup tasks. This makes sure that idle gets scheduled and the watchdog feed.

Ah yes that makes sense. Small potential improvement: use `delay(0)`. At least on ESP8266s that's often used for a similar reason.

> Or we indeed just tell the task watchdog that a higher "idle task starvation" is ok during startup phase by setting the timeout to e.g. 10s (what the code is currently doing).

The problem I see there is what constant is enough? If the idle task needs to be scheduled, how do we know that 10s will work for all cases?

agners replied:
> Ah yes that makes sense. Small potential improvement: use delay(0). At least on ESP8266s that's often used for a similar reason.

That seems not to work in this case. I assume that `0` really does not lead to idle task getting scheduled, at least in that case.

> The problem I see there is what constant is enough? If the idle task needs to be scheduled, how do we know that 10s will work for all cases?

The majority of time is really spent waiting for WiFi scan and connect, which seems to be around 4s. I think 10s would be fairly safe.

That said, I kinda like the `delay(1)` solution better: This is a state where we indeed wait for something asynchronously to happen. Letting the CPU go idle for one tick (scheduler runs at 1000Hz it seems) doesn't seem a bad thing.

agners replied:
New findings: The Task Watchdog is **not** active on the CPU which Arduino runs on dual-core ESP32! 

This can be shown by adding a `while(true);` in a `loop()` function: It does not trigger the a watchdog! The behavior does make sense after studying [sdkconfig.h](https://github.com/espressif/arduino-esp32/blob/2.0.0-rc1/tools/sdk/esp32/include/config/sdkconfig.h) more closely: `CONFIG_TASK_WDT_CHECK_IDLE_TASK_CPU0` is set to `1`, but also `CONFIG_ARDUINO_RUNNING_CORE` is set to `1`. So in other words. it seems on dual-core ESP32, we only monitor CPU0 (where WiFi and some other tasks run).

This is also not new to 2.0.0-rc1, at least Arduino 1.0.4 seems to have the same configurations in [sdkconfig.h](https://github.com/espressif/arduino-esp32/blob/1.0.4/tools/sdk/include/config/sdkconfig.h).

Enabling Task Watchdog does show the same on the dual-core

```
void Application::setup() {
  TaskHandle_t idle_1 = xTaskGetIdleTaskHandleForCPU(1);
  esp_task_wdt_init(2, false);
  esp_task_wdt_add(idle_1);
```
(I had to lower the timeout to 2s, as the whole `setup()` process is faster on dual-core ESP32 variants)

```
[12:48:17][I][wifi:254]: WiFi Connecting to '<SSID>'...
[12:48:17][V][wifi:256]: Connection Params:
[12:48:17][V][wifi:257]:   SSID: '<SSID>'
[12:48:17][V][wifi:260]:   BSE (4672) task_wdt: Task watchdog got triggered. The following tasks did not reset the watchdog in time:
[12:48:17]E (4672) task_wdt:  - IDLE (CPU 1)
[12:48:17]E (4672) task_wdt: Tasks currently running:
[12:48:17]E (4672) task_wdt: CPU 0: IDLE
[12:48:17]E (4672) task_wdt: CPU 1: loopTask
```

Ideally I think we should enable also a watchdog for ESP32.  The above code should do this.

For ESP32-C3, I'll use `delay(1)` to make sure watchdog feeding work.

OttoWinter replied:
> Ideally I think we should enable also a watchdog for ESP32. The above code should do this.

Oh didn't know that, yes that is a very good idea. Thanks!

agners committed:
Revert "Increase task watchdog timeout during setup()"

This reverts commit 6a7ac4b7badbd8ed89105adcc49c54c5b58d9ac2.

Try a different approach by making sure watchdog actually getting
triggered.

agners committed:
Make sure to feed the watchdog during setup()

Use feed_wdt() during setup to avoid watchdog timeouts during setup.

agners committed:
Fix Watchdog reset on ESP32

For ESP32 a Task Watchdog is used. However, yield() doesn't trigger the
task watchdog. The Task Watchdog is attached to the idle task, basically
making sure that the CPU isn't overloaded and goes into idle state at
least once every 5s. To force a Watchdog feed, we need to use a delay(1).
